### PR TITLE
Ability to handle SniHandler reference count leak

### LINKED ISSUE #5678: SniHandler#replaceHandler() may leak (reference counted) SslContexts
The SniHandler#select()/AsyncMapper#map() methods execute asynchronously and by the time the returned Future completes the client is possibly no longer connected.

The #selection field will be set but probably after #channelInactive() (i.e. can't really do anything with it) and the #replace() call on the ChannelPipeline will probably throw a NoSuchElementException.

I have no good suggestion how to fix it. There is an opportunity to cancel the future and it could try to decrement the refCnt if the Future completes successfully but the client has disconnected in the meantime. 

Given SniHandler is a one-shot handler maybe something along the lines of this?!

```java
class SniHandler {
  private Promsie<SslContext> promise;

  void handlerAdded(ChannelHandlerContext ctx) {
    promsie = ctx.newPromise();
  }

  void channelInactive(ChannelHandlerContext ctx) {
    promsie.cancel(true);
    promsie.addFutureListener(... release() on success ...);
    ctx.fireChannelInactive();
  }

  void select(...) {
    mapping.map(hostname, promsie).addFutureListener(... replaceHandler() ...);
  }
}
```

### PR DESCRIPTION
**Pull request by rkapsi**
Motivation:
The SniHandler attempts to wrap the selected SslContext in a SslHandler and insert that SslHandler into the pipeline. However if the underlying channel has been closed or the pipeline has been modified the pipeline.replace(..) operation may fail and we will leak the SslContext if it's reference counted.

Modifications:

Add a protected replaceHandler(...) method to SniHandler that users can override to implement their custom behavior.

Result:

Fixes #5678 by giving the user the ability to override replaceHandler(...) and handle errors and possibly release the SslContext.

This change facilitates addtionally reference counting from the user's POV as they now have a way to get their hands onto the SslContext and maybe insert an another ChannelHandler into the pipeline (along to SslHandler) that will -- the reference count on channelInactive().

The SslHandler has also a few setter methods such as setHandshakeTimeout() that were previously inaccessible with the SniHandler. The user can now implement their own replaceHandler() version.

### CODE REVIEW

**Comment by rkapsi** [comment]
@Scottmitch PTAL

File: handler/src/main/java/io/netty/handler/ssl/SniHandler.java [hunk 1]
```diff
@@ -303,8 +303,24 @@ public void operationComplete(Future<SslContext> future) throws Exception {
...
 
     private void replaceHandler(ChannelHandlerContext ctx, Selection selection) {
-        this.selection = selection;
-        SslHandler sslHandler = selection.context.newHandler(ctx.alloc());
+        try {
+            this.selection = selection;
+            replaceHandler(ctx, selection.hostname, selection.context);
+        } catch (Throwable cause) {
+            this.selection = EMPTY_SELECTION;
+            ctx.fireExceptionCaught(cause);
```
Comment (rkapsi) [inline handler/src/main/java/io/netty/handler/ssl/SniHandler.java]: Maybe even fire a custom `SniHandlerException(hostname, context, cause)` to give the user every possible opportunity not to lose track of the SslContext.

File: handler/src/main/java/io/netty/handler/ssl/SniHandler.java [hunk 2]
```diff
@@ -303,8 +303,24 @@ public void operationComplete(Future<SslContext> future) throws Exception {
...
+    }
+
+    /**
+     * Replaces the {@link SniHandler} with a {@link SslHandler} as provided by the
+     * given {@link SslContext}.
+     *
+     * Users may override this method to implement custom behavior and to get a
+     * hold onto the {@link SslContext}.
+     */
+    protected void replaceHandler(ChannelHandlerContext ctx, String hostname, SslContext context) throws Exception {
```
Comment (netkins) [inline handler/src/main/java/io/netty/handler/ssl/SniHandler.java]: MAJOR Remove this unused method parameter "hostname". [rule](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1172)

File: handler/src/main/java/io/netty/handler/ssl/SniHandler.java (hunk 1 above)
Comment (Scottmitch) [inline handler/src/main/java/io/netty/handler/ssl/SniHandler.java]: Firing exceptions with reference counted objects can create confusion in my experience. IMO lets avoid this and push folks toward `replaceHandler`.

**Comment by Scottmitch** [comment]
@rkapsi - lets discuss the reference count issue in https://github.com/netty/netty/pull/5745#issuecomment-242458567 ... we may consider extensibility as an improvement but I'm not sure it should be used to address this reference counting issue.

**Comment by rkapsi** [comment]
Closing and opening a new one later.

**Comment by Scottmitch** [comment]
thanks @rkapsi !

**Pull request closed without merging**